src/nrniv/memory_usage.cpp

```diff
@@ -0,0 +1,157 @@
+#include <neuron/model_data.hpp>
+#include <neuron/cache/model_data.hpp>
+#include <stdexcept>
+#include "oc_ansi.h"
+
+namespace neuron::container {
+ModelMemoryUsage memory_usage(const Model& model) {
+    auto nodes = memory_usage(model.node_data());
+
+    auto mechanisms = StorageMemoryUsage();
+    model.apply_to_mechanisms([&mechanisms](const auto& md) { mechanisms += memory_usage(md); });
+
+    return {nodes, mechanisms};
+}
+
+cache::ModelMemoryUsage memory_usage(const std::optional<neuron::cache::Model>& model) {
+    return model ? memory_usage(*model) : cache::ModelMemoryUsage{};
+}
+cache::ModelMemoryUsage memory_usage(const neuron::cache::Model& model) {
```

Comment:
Wouldn't it be safer if there was a member function for `neuron::cache::Model` that computed the memory usage and was called from here instead of doing all this work from here? Something similar may apply for other classes as well.
My rationale is that in case we add a new member variable in the `neuron::cache::Model` class for example we will have to remember to come here and add it to the computation. Instead if there was a member function in this class it would be harder to miss

Reply:
Yes, though I think it's more important to merge and reduce the custom setup required to do our production runs. (Which currently require picking just the right commits from several branches.)